ebos/ecloutputblackoilmodule.hh

```diff
@@ -1038,6 +1107,304 @@ public:
         }
 
     }
+	
+	// write production report to output 
+    void outputProdLog(size_t reportStepNum, const bool substep, bool forceDisableProdOutput) 
+    {
+		if (!substep) {
+			
+			ScalarBuffer  tmp_values(WellProdDataType::numWPValues, 0.0);
+			StringBuffer  tmp_names(WellProdDataType::numWPNames, "");
+			outputProductionReport_(tmp_values, tmp_names, forceDisableProdOutput);
+			
+			const auto& st = simulator_.vanguard().summaryState(); 
+			const auto& schedule = simulator_.vanguard().schedule();
+			for (const auto& well: schedule.getWells2(reportStepNum)) {
+				
+				// don't bother with wells not on this process 
+				const auto& defunctWellNames = simulator_.vanguard().defunctWellNames(); 
+				if (defunctWellNames.find(well.name()) != defunctWellNames.end()) {
+					continue;
+				}				
+					
+				// Ignore injector wells  
+				if (well.isInjector()){ 
+					continue;
+				}
+
+			    std::ostringstream wellName; 
+				wellName << well.name(); 	
+				tmp_names[0] = wellName.str();//WellProdDataType::WellName
+	
+				
+				auto wName = static_cast<std::string>(well.name());                
+                auto get = [&st, &wName](const std::string& vector)
+                {
+                const auto key = vector + ':' + wName;
+
+                return st.has(key) ? st.get(key) : 0.0;
+                };
+				
+				const auto& controls = well.productionControls(st);
+				using CMode = ::Opm::Well2::ProducerCMode;
+				
+				auto fctl = [](const auto wmctl)
+				{
+					switch (wmctl) {
+					case CMode::ORAT: return "OilRate";
+					case CMode::WRAT: return "WatRate";
+					case CMode::GRAT: return "GasRate";
+					case CMode::LRAT: return "LiqRate";
+					case CMode::RESV: return "ResVRate";
+					case CMode::THP:  return "THP";
+					case CMode::BHP:  return "BHP";
+					case CMode::CRAT: return "CombRate";
+					case CMode::GRUP: return "Group";
```

Comment:
You should use the abbreviations used in the manual unless the commercial simulator print files has the full name. Nearly all reservoir engineers know this nomenclature, it has been around from the 1970s. It is important to match the report format as closely as possible, as engineers use the reports to scrape data, although that is becoming less common these days.
```

                                                          PRODUCTION REPORT
                                                          -----------------

------------------------------------------------------------------------------------------------------------------------------------
:  WELL  :  LOCATION :CTRL:    OIL    :   WATER   :    GAS    :   FLUID   :WATER: GAS/OIL: WAT/GAS: BHP OR : THP OR : STEADY-ST PI :
:  NAME  :  (I,J,K)  :MODE:    RATE   :   RATE    :    RATE   :  RES.VOL. : CUT :  RATIO :  RATIO : CON.PR.: BLK.PR.: OR POTN OF   :
: OR GRID:           :    :   SM3/DAY :   SM3/DAY :   SM3/DAY :   RM3/DAY :     : SM3/SM3: SM3/SM3:  BARSA :  BARSA : PREFERRED PH :
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:FIELD               :    :    17040.1:       18.6:  1826446.6:    22322.8:0.001:  107.19:  0.0000:        :    20.0:              :
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:INJE       GRP LV  1:    :        0.0:        0.0:        0.0:        0.0:0.000:    0.00:  0.0000:        :        :              :
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:MANI-C     GRP LV  2:    :        0.0:        0.0:        0.0:        0.0:0.000:    0.00:  0.0000:        :        :              :
------------------------------------------------------------------------------------------------------------------------------------
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:MANI-F     GRP LV  2:    :        0.0:        0.0:        0.0:        0.0:0.000:    0.00:  0.0000:        :        :              :
------------------------------------------------------------------------------------------------------------------------------------
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:WI-GSEG    GRP LV  2:    :        0.0:        0.0:        0.0:        0.0:0.000:    0.00:  0.0000:        :        :              :
------------------------------------------------------------------------------------------------------------------------------------
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:PROD       GRP LV  1:    :    17040.1:       18.6:  1826446.6:    22322.8:0.001:  107.19:  0.0000:        :    20.0:              :
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:MANI-B2    GRP LV  2:    :        0.0:        0.0:        0.0:        0.0:0.000:    0.00:  0.0000:        :    58.1:              :
------------------------------------------------------------------------------------------------------------------------------------
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:MANI-B1    GRP LV  2:    :     5247.8:        2.7:   567180.0:     6885.7:0.001:  108.08:  0.0000:        :    39.9:              :
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:MANI-K1    GRP LV  3:    :        0.0:        0.0:        0.0:        0.0:0.000:    0.00:  0.0000:        :    39.9:              :
------------------------------------------------------------------------------------------------------------------------------------
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:B1-DUMMY   GRP LV  3:    :     5247.8:        2.7:   567180.0:     6885.7:0.001:  108.08:  0.0000:        :    39.9:              :
------------------------------------------------------------------------------------------------------------------------------------
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:B-2H    : 15, 31    :RESV:     5247.8:        2.7:   567180.0:     6885.7:0.001:  108.08:  0.0000:   228.9:     0.0:   274.4  O PI:
:  BLOCK : 17, 31,  9:    :      486.4:        0.0:    52676.2:      638.4:0.000:  108.31:  0.0000:   228.9:   251.7:    21.3  O PI:
:  BLOCK : 19, 31,  9:    :      285.5:        1.0:    30885.4:      375.6:0.003:  108.19:  0.0000:   229.0:   250.3:    13.4  O PI:
:  BLOCK : 20, 31, 10:    :      861.0:        0.5:    92884.6:     1129.2:0.001:  107.87:  0.0000:   229.7:   249.6:    43.3  O PI:
:  BLOCK : 21, 31, 10:    :      285.2:        0.1:    30775.1:      374.0:0.001:  107.91:  0.0000:   229.7:   249.5:    14.4  O PI:
:  BLOCK : 22, 32, 10:    :     1462.5:        0.6:   158043.7:     1918.7:0.000:  108.06:  0.0000:   229.4:   247.4:    81.6  O PI:
:  BLOCK : 24, 32, 10:    :      911.5:        0.2:    98583.8:     1196.1:0.000:  108.15:  0.0000:   229.3:   246.9:    51.6  O PI:
:  BLOCK : 25, 32, 10:    :      640.5:        0.1:    69281.8:      840.5:0.000:  108.16:  0.0000:   229.3:   247.0:    36.2  O PI:
:  BLOCK : 29, 33, 10:    :      315.1:        0.1:    34049.5:      413.4:0.000:  108.06:  0.0000:   229.5:   254.6:    12.6  O PI:
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:MANI-D1    GRP LV  2:    :     5713.6:        7.5:   607821.1:     7472.7:0.001:  106.38:  0.0000:        :    41.1:              :
------------------------------------------------------------------------------------------------------------------------------------
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:D-1H    : 22, 22    :RESV:     5713.6:        7.5:   607821.1:     7472.7:0.001:  106.38:  0.0000:   174.6:     0.0:    94.8  O PI:
:  BLOCK : 22, 22,  5:    :      240.7:        1.9:    25401.3:      315.9:0.008:  105.54:  0.0001:   174.6:   235.6:     3.9  O PI:
:  BLOCK : 23, 22,  6:    :        4.4:        0.0:      467.8:        5.8:0.008:  106.08:  0.0001:   174.7:   236.2:     0.1  O PI:
:  BLOCK : 23, 22,  7:    :      212.1:        1.9:    22104.4:      277.8:0.009:  104.21:  0.0001:   175.1:   235.4:     3.5  O PI:
:  BLOCK : 23, 22,  9:    :      850.1:        2.7:    91149.1:     1115.5:0.003:  107.22:  0.0000:   175.9:   233.3:    14.8  O PI:
:  BLOCK : 23, 22, 10:    :     2789.4:        0.0:   297617.4:     3646.9:0.000:  106.70:  0.0000:   176.6:   235.6:    47.3  O PI:
:  BLOCK : 23, 22, 11:    :      629.7:        0.0:    65717.4:      819.2:0.000:  104.36:  0.0000:   177.3:   239.4:    10.1  O PI:
:  BLOCK : 23, 22, 12:    :       36.6:        0.0:     3876.1:       47.8:0.000:  105.81:  0.0000:   177.6:   242.1:     0.6  O PI:
:  BLOCK : 23, 22, 13:    :      950.6:        0.9:   101487.5:     1243.9:0.001:  106.76:  0.0000:   178.1:   243.9:    14.5  O PI:
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:MANI-D2    GRP LV  2:    :     6078.7:        8.4:   651445.5:     7964.4:0.001:  107.17:  0.0000:        :    42.3:              :
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:MANI-K2    GRP LV  3:    :        0.0:        0.0:        0.0:        0.0:0.000:    0.00:  0.0000:        :    42.3:              :
------------------------------------------------------------------------------------------------------------------------------------
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:D2-DUMMY   GRP LV  3:    :     6078.7:        8.4:   651445.5:     7964.4:0.001:  107.17:  0.0000:        :    42.3:              :
------------------------------------------------------------------------------------------------------------------------------------
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:D-2H    : 14, 28    :RESV:     6078.7:        8.4:   651445.5:     7964.4:0.001:  107.17:  0.0000:   229.7:     0.0:   479.2  O PI:
:  BLOCK : 14, 26,  9:    :      538.2:        0.0:    58797.1:      708.9:0.000:  109.26:  0.0000:   229.7:   250.0:    26.6  O PI:
:  BLOCK : 14, 25,  9:    :      895.0:        0.0:    97478.4:     1177.4:0.000:  108.91:  0.0000:   229.6:   248.0:    48.8  O PI:
:  BLOCK : 14, 23,  9:    :      210.8:        0.0:    22811.5:      276.6:0.000:  108.20:  0.0000:   229.8:   246.6:    12.5  O PI:
:  BLOCK : 14, 22,  9:    :     1493.7:        0.0:   160236.7:     1955.3:0.000:  107.28:  0.0000:   229.9:   241.9:   123.8  O PI:
:  BLOCK : 14, 21,  9:    :     1117.1:        2.9:   118884.4:     1462.7:0.003:  106.42:  0.0000:   230.0:   239.5:   118.4  O PI:
:  BLOCK : 14, 20,  9:    :      306.1:        0.8:    32535.5:      400.7:0.003:  106.31:  0.0000:   230.2:   243.1:    23.8  O PI:
:  BLOCK : 14, 15,  9:    :      328.3:        1.2:    34693.1:      429.5:0.004:  105.67:  0.0000:   230.2:   241.9:    28.1  O PI:
:  BLOCK : 14, 14,  9:    :      823.6:        2.5:    86762.0:     1076.2:0.003:  105.35:  0.0000:   229.9:   241.3:    72.9  O PI:
:  BLOCK : 14, 13,  9:    :      366.0:        1.0:    39246.8:      480.2:0.003:  107.22:  0.0000:   229.8:   244.8:    24.4  O PI:
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:MANI-E1    GRP LV  2:    :        0.0:        0.0:        0.0:        0.0:0.000:    0.00:  0.0000:        :    55.2:              :
------------------------------------------------------------------------------------------------------------------------------------
====================================================================================================================================
:        :           :    :           :           :           :           :     :        :        :        :        :              :
:MANI-E2    GRP LV  2:    :        0.0:        0.0:        0.0:        0.0:0.000:    0.00:  0.0000:        :    55.2:              :
------------------------------------------------------------------------------------------------------------------------------------
====================================================================================================================================

```

Reply:
P.S. As an RE I think it is fantastic that you are implementing these reports, it will make the program so much more user friendly for the engineers. So thank you. 

I cannot code C++, but if you need help with anything else I'm more than willing to help out.

Reply:
Thank you for your insight. I will output as the commercial simulators.